### PR TITLE
Log outgoing `PrimitiveSet` protos 

### LINKED ISSUE #1745: Logging for outgoing PrimitiveSet protos
### Description of the task

We want to log outgoing primitive protos to allow for validation of behaviour with replayed real world data.

### Acceptance criteria

- [ ] Outgoing `PrimitiveSet` messages are logged to files

### PR DESCRIPTION
**Pull request by jordanzliu**
### Description

Log outgoing `PrimitiveSet` protos to a directory structure such as: 
- `OutputDir`
-- `SensorMsg`
-- `PrimitiveSet`

### Testing Done

Subbed in new `ProtoLogger` for `ReplayLogger` and repurposed old replay tests

### Resolved Issues

Resolves #1745 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/full_system_main.cpp [hunk 1]
```diff
@@ -94,11 +95,29 @@ int main(int argc, char **argv)
             backend->Subject<SensorProto>::registerObserver(visualizer);
         }
 
-        if (!args->replay_output_dir()->value().empty())
+        // TODO: manage this mess better and don't log SensorMsg when we are replaying
+        // TODO: make a shim between SensorFusion and ProtoLogger to log filtered world
```
Comment (jonathanlew) [inline src/software/full_system_main.cpp]: sorry what would this shim do?

File: src/software/proto/proto_logger/proto_log_reader.h [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+    size_t max_chunk_idx;
+    size_t cur_chunk_idx;
+    int cur_msg_idx;
+    RepeatedAnyMsg cur_chunk;
+    std::experimental::filesystem::path replay_dir;
+};
+
+template <typename Msg>
+std::optional<Msg> ProtoLogReader::getNextMsg()
```
Comment (jonathanlew) [inline src/software/proto/proto_logger/proto_log_reader.h]: is the idea to use one `ProtoLogReader` to read multiple message types? If not, then it would more consistent to make it a template class

File: src/software/proto/repeated_any_msg.proto [hunk 3]
```diff
@@ -0,0 +1,8 @@
+syntax = 'proto3';
+import "google/protobuf/any.proto";
+
+message RepeatedAnyMsg
+{
+    string message_type                   = 1;
+    repeated google.protobuf.Any messages = 2;
```
Comment (jonathanlew) [inline src/software/proto/repeated_any_msg.proto:7]: have you considered: https://developers.google.com/protocol-buffers/docs/techniques#self-description

File: src/software/proto/proto_logger/proto_logger.tpp [hunk 4]
```diff
@@ -0,0 +1,106 @@
+#include <google/protobuf/util/delimited_message_util.h>
+
+#include <fstream>
+
+#include "software/logger/logger.h"
+#include "software/proto/proto_logger/proto_logger.h"
+
+template <typename Msg>
```
Comment (jonathanlew) [inline src/software/proto/proto_logger/proto_logger.tpp]:
```suggestion
template <typename MSG_TYPE>
```
for clarity that it's a type

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Got a few comments, but I haven't tried it yet

File: src/software/full_system_main.cpp (hunk 1 above)
Comment (jordanzliu) [reply src/software/full_system_main.cpp]: it would be an `Observer<World>` and `Subject<TbotsProto::Vision>` and do the conversion

File: src/software/proto/proto_logger/proto_log_reader.h (hunk 2 above)
Comment (jordanzliu) [reply src/software/proto/proto_logger/proto_log_reader.h]: I'm not a big fan of making things templates unless necessary, for the sake of compile times. In this class, this is the only method that requires the type of the message to be converted into, so only this method was templated.

File: src/software/proto/repeated_any_msg.proto (hunk 3 above)
Comment (jordanzliu) [reply src/software/proto/repeated_any_msg.proto:7]: This might be worthwhile if we're going to build a Python environment, will look into it

File: src/software/full_system_main.cpp (hunk 1 above)
Comment (jonathanlew) [reply src/software/full_system_main.cpp]: That's pretty close to what a `Backend` does

Comment (jordanzliu) [reply src/software/full_system_main.cpp]: Should we have backend connect `Subject`s to `ProtoLogger`s? Right now, only `full_system_main.cpp` is responsible for the "plumbing" between components. Alternatively, maybe we could have a global singleton to manage protobuf logging, I'll give that a try.

Comment (jonathanlew) [reply src/software/full_system_main.cpp]: It would be good idea to take the time to plan out an architecture before diving right in

Comment (jordanzliu) [reply src/software/full_system_main.cpp]: Rescoped the PR to only concern logging `PrimitiveSet`, logging filtered world data is gonna take some more design. Opened #1789 to track this

File: src/software/proto/proto_logger/proto_logger.tpp (hunk 4 above)
Comment (jordanzliu) [reply src/software/proto/proto_logger/proto_logger.tpp]: renamed

File: src/software/proto/repeated_any_msg.proto (hunk 3 above)
Comment (jordanzliu) [reply src/software/proto/repeated_any_msg.proto:7]: Out of scope for now, will consider it for the python environment proposal. I don't think working with descriptor protos is well-supported for Python though, the linked page says `DynamicMessage` is only for C++ and Java.

File: src/software/multithreading/threaded_observer.tpp [hunk 5]
```diff
@@ -1,6 +1,7 @@
 #pragma once
 
 #include "software/multithreading/threaded_observer.h"
+#include "software/util/typename/typename.h"
```
Comment (jonathanlew) [inline src/software/multithreading/threaded_observer.tpp]: nit: why was this added?

**Review by jonathanlew — COMMENTED** [review]
nit about typename names. Thoughts on TMsg? https://stackoverflow.com/questions/692752/naming-conventions-for-template-types

otherwise looks good

**Comment by jordanzliu** [comment]
there's `PlannerT` in our codebase already, and theres also some all-caps typenames in the codebase. I'll leave `MsgT`, but we can discuss the convention for the entire codebase at another time.

File: src/software/multithreading/threaded_observer.tpp (hunk 5 above)
Comment (jordanzliu) [reply src/software/multithreading/threaded_observer.tpp]: oops, fixed

**Review by jonathanlew — APPROVED** [review]

**Comment by akhilveeraghanta** [comment]
@jordanzliu how do I run this locally?

`bazel run //software:full_system -- --backend=ReplayBackend --proto_log_output_dir=/tmp` fails with
~~~
terminate called after throwing an instance of 'std::invalid_argument'
  what():  does not exist or is not a directory!
2020/11/01 08:59:39 687691

***** FATAL SIGNAL RECEIVED ******* 
Received fatal signal: SIGABRT(6)	PID: 1256073
~~~

**Comment by jordanzliu** [comment]
`ReplayBackend` must be used with `--replay_input_dir <directory>`

**Review by akhilveeraghanta — APPROVED** [review]
Ah rookie mistake on my end, LGTM :fire:
